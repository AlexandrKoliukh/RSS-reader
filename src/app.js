import { watch } from 'melanke-watchjs';
import { getRssData, checkInputValid } from './controllers';
import {
  renderForm, renderItems, renderStreams, valid,
} from './watchers';

export default () => {
  const state = {
    fetchedUrls: [],
    fetchingState: 'none',
    rssItems: [],
    rssStreams: [],
    isValid: false,
    url: '',
  };

  const rssUrlInput = document.getElementById('rssUrl');
  const mainForm = document.getElementById('mainForm');

  rssUrlInput.addEventListener('input', e => checkInputValid(e, state));
  rssUrlInput.addEventListener('paste', e => checkInputValid(e, state));

  mainForm.addEventListener('reset', () => {
    rssUrlInput.dispatchEvent(new Event('input'));
  });
  mainForm.addEventListener('submit', e => getRssData(e, state));

  watch(state, 'fetchingState', () => renderForm(state));
  watch(state, 'rssItems', () => renderItems(state));
  watch(state, 'rssStreams', () => renderStreams(state));
  watch(state, 'isValid', () => valid(state));
};
