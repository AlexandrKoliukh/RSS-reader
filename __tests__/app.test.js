// @ts-check
import 'core-js/stable';
import 'regenerator-runtime/runtime';
import { promises as fs } from 'fs';
import path from 'path';
import { screen, waitFor } from '@testing-library/dom';
import userEvent from '@testing-library/user-event';
import prettier from 'prettier';
import nock from 'nock';
import MutationObserver from 'mutation-observer';
import axios from 'axios';
import httpAdapter from 'axios/lib/adapters/http';
import app from '../src/app';

global.MutationObserver = MutationObserver;
axios.defaults.adapter = httpAdapter;
nock.disableNetConnect();


const prettierOptions = {
  parser: 'html',
  htmlWhitespaceSensitivity: 'ignore',
  tabWidth: 4,
};
const getTree = (el) => prettier.format(el.innerHTML, prettierOptions);
const testUrl = 'http://lorem-rss.herokuapp.com/feed';
const getFixturePath = (fileName) => path.resolve(__dirname, `../__fixtures__/${fileName}`);

let scope;

beforeEach(async () => {
  const pathToHtml = getFixturePath('index.html');
  const rssPath = getFixturePath('rss.xml');
  const html = await fs.readFile(pathToHtml, 'utf8');
  const rssData = await fs.readFile(rssPath, 'utf8');

  scope = nock('https://api.allorigins.win')
    .get('/raw')
    .query({ url: testUrl })
    .reply(200, rssData);

  document.body.innerHTML = html;
  app();
});

afterEach(async () => {
  nock.cleanAll();
});

test('initial state', async () => {
  const list = screen.getByTestId('rssList');
  const streamList = screen.getByTestId('streamList');
  const form = screen.getByTestId('urlForm');
  expect(getTree(list)).toMatchSnapshot();
  expect(getTree(streamList)).toMatchSnapshot();
  expect(getTree(form)).toMatchSnapshot();
});

test('Fetch and render RSS', async () => {
  const urlInput = screen.getByTestId('urlInput');
  const form = screen.getByTestId('urlForm');
  const list = screen.getByTestId('rssList');
  // const urlSubmit = document.getElementById('urlSubmit');
  const streamList = screen.getByTestId('streamList');

  userEvent.type(urlInput, testUrl);
  // userEvent.click(urlSubmit);
  form.dispatchEvent(new Event('submit'));

  await waitFor(() => {
    scope.done();
  });

  const showBtn = document.querySelector('[data-target="#modal"]');

  const modal = screen.getByTestId('modalContainer');

  userEvent.click(showBtn);

  expect(urlInput.value).toBe('');
  expect(getTree(list)).toMatchSnapshot();
  expect(getTree(streamList)).toMatchSnapshot();
  expect(getTree(modal)).toMatchSnapshot();
});
