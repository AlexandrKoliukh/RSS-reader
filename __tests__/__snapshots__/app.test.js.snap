// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fetch and render RSS 1`] = `
"<div class=\\"card\\">
    <a href=\\"http://example.com/test/1607584920\\">
        Lorem ipsum 2020-12-10T07:22:00Z
    </a>
    <button
        type=\\"button\\"
        class=\\"btn btn-primary\\"
        data-toggle=\\"modal\\"
        data-target=\\"#modal\\"
        data-href=\\"http://example.com/test/1607584920\\"
    >
        Показать больше
    </button>
</div>
<div class=\\"card\\">
    <a href=\\"http://example.com/test/1607584860\\">
        Lorem ipsum 2020-12-10T07:21:00Z
    </a>
    <button
        type=\\"button\\"
        class=\\"btn btn-primary\\"
        data-toggle=\\"modal\\"
        data-target=\\"#modal\\"
        data-href=\\"http://example.com/test/1607584860\\"
    >
        Показать больше
    </button>
</div>
<div class=\\"card\\">
    <a href=\\"http://example.com/test/1607584800\\">
        Lorem ipsum 2020-12-10T07:20:00Z
    </a>
    <button
        type=\\"button\\"
        class=\\"btn btn-primary\\"
        data-toggle=\\"modal\\"
        data-target=\\"#modal\\"
        data-href=\\"http://example.com/test/1607584800\\"
    >
        Показать больше
    </button>
</div>
<div class=\\"card\\">
    <a href=\\"http://example.com/test/1607584740\\">
        Lorem ipsum 2020-12-10T07:19:00Z
    </a>
    <button
        type=\\"button\\"
        class=\\"btn btn-primary\\"
        data-toggle=\\"modal\\"
        data-target=\\"#modal\\"
        data-href=\\"http://example.com/test/1607584740\\"
    >
        Показать больше
    </button>
</div>
<div class=\\"card\\">
    <a href=\\"http://example.com/test/1607584680\\">
        Lorem ipsum 2020-12-10T07:18:00Z
    </a>
    <button
        type=\\"button\\"
        class=\\"btn btn-primary\\"
        data-toggle=\\"modal\\"
        data-target=\\"#modal\\"
        data-href=\\"http://example.com/test/1607584680\\"
    >
        Показать больше
    </button>
</div>
<div class=\\"card\\">
    <a href=\\"http://example.com/test/1607584620\\">
        Lorem ipsum 2020-12-10T07:17:00Z
    </a>
    <button
        type=\\"button\\"
        class=\\"btn btn-primary\\"
        data-toggle=\\"modal\\"
        data-target=\\"#modal\\"
        data-href=\\"http://example.com/test/1607584620\\"
    >
        Показать больше
    </button>
</div>
<div class=\\"card\\">
    <a href=\\"http://example.com/test/1607584560\\">
        Lorem ipsum 2020-12-10T07:16:00Z
    </a>
    <button
        type=\\"button\\"
        class=\\"btn btn-primary\\"
        data-toggle=\\"modal\\"
        data-target=\\"#modal\\"
        data-href=\\"http://example.com/test/1607584560\\"
    >
        Показать больше
    </button>
</div>
<div class=\\"card\\">
    <a href=\\"http://example.com/test/1607584500\\">
        Lorem ipsum 2020-12-10T07:15:00Z
    </a>
    <button
        type=\\"button\\"
        class=\\"btn btn-primary\\"
        data-toggle=\\"modal\\"
        data-target=\\"#modal\\"
        data-href=\\"http://example.com/test/1607584500\\"
    >
        Показать больше
    </button>
</div>
<div class=\\"card\\">
    <a href=\\"http://example.com/test/1607584440\\">
        Lorem ipsum 2020-12-10T07:14:00Z
    </a>
    <button
        type=\\"button\\"
        class=\\"btn btn-primary\\"
        data-toggle=\\"modal\\"
        data-target=\\"#modal\\"
        data-href=\\"http://example.com/test/1607584440\\"
    >
        Показать больше
    </button>
</div>
<div class=\\"card\\">
    <a href=\\"http://example.com/test/1607584380\\">
        Lorem ipsum 2020-12-10T07:13:00Z
    </a>
    <button
        type=\\"button\\"
        class=\\"btn btn-primary\\"
        data-toggle=\\"modal\\"
        data-target=\\"#modal\\"
        data-href=\\"http://example.com/test/1607584380\\"
    >
        Показать больше
    </button>
</div>
"
`;

exports[`Fetch and render RSS 2`] = `
"<div class=\\"card\\">
    <h4>This is a constantly updating lorem ipsum feed</h4>
    <p>Lorem ipsum feed for an interval of 1 minutes with 10 item(s)</p>
</div>
"
`;

exports[`Fetch and render RSS 3`] = `
"<div class=\\"modal-dialog\\" role=\\"document\\">
    <div class=\\"modal-content\\">
        <div class=\\"modal-header\\">
            <h5 class=\\"modal-title\\" id=\\"modalTitle\\">
                Lorem ipsum 2020-12-10T07:22:00Z
            </h5>
            <button
                type=\\"button\\"
                class=\\"close\\"
                data-dismiss=\\"modal\\"
                aria-label=\\"Close\\"
            >
                <span aria-hidden=\\"true\\">×</span>
            </button>
        </div>
        <div class=\\"modal-body\\" id=\\"modalBody\\">
            Anim do culpa laboris Lorem aliqua aute consectetur cupidatat cillum
            amet aliqua occaecat.
        </div>
        <div class=\\"modal-footer\\">
            <button
                type=\\"button\\"
                class=\\"btn btn-secondary\\"
                data-dismiss=\\"modal\\"
            >
                Close
            </button>
        </div>
    </div>
</div>
"
`;

exports[`initial state 1`] = `""`;

exports[`initial state 2`] = `""`;

exports[`initial state 3`] = `
"<label for=\\"rssUrl\\">Введите адрес:</label>
<input
    type=\\"text\\"
    class=\\"form-control border\\"
    id=\\"rssUrl\\"
    data-testid=\\"urlInput\\"
/>
<small id=\\"rssUrlHelp\\" class=\\"form-text text-muted\\">
    Пример: http://lorem-rss.herokuapp.com/feed
</small>
<input
    type=\\"submit\\"
    disabled=\\"\\"
    id=\\"urlSubmit\\"
    class=\\"form-control btn btn-primary mt-2\\"
    data-testid=\\"urlSubmit\\"
/>
"
`;
